clients/drcachesim/tracer/raw2trace.cpp

```diff
@@ -1032,25 +1036,89 @@ raw2trace_t::process_syscall_pt(raw2trace_thread_data_t *tdata, uint64_t syscall
     }
     trace_entry_t end_entry = { .type = TRACE_TYPE_MARKER,
                                 .size = TRACE_MARKER_TYPE_SYSCALL_TRACE_END,
-                                .addr = 0 };
+                                .addr = sysnum };
     entries.push_back(end_entry);
     if (entries.size() == 2) {
         tdata->error = "No trace entries generated from PT data";
         return false;
     }
 
     accumulate_to_statistic(tdata, RAW2TRACE_STAT_SYSCALL_TRACES_DECODED, 1);
+    app_pc saved_decode_pc;
+    trace_entry_t entries_with_encodings[WRITE_BUFFER_SIZE];
+    trace_entry_t *buf = entries_with_encodings;
     for (const auto &entry : entries) {
-        if (type_is_instr(static_cast<trace_type_t>(entry.type)))
+        if (type_is_instr(static_cast<trace_type_t>(entry.type))) {
+            if (buf != entries_with_encodings) {
+                if (!write(tdata, entries_with_encodings, buf, &saved_decode_pc, 1)) {
+                    return false;
+                }
+                buf = entries_with_encodings;
+            }
+            app_pc instr_pc = reinterpret_cast<app_pc>(entry.addr);
             accumulate_to_statistic(tdata, RAW2TRACE_STAT_KERNEL_INSTR_COUNT, 1);
+            if (tdata->syscall_pc_to_decode_pc_.find(instr_pc) ==
+                tdata->syscall_pc_to_decode_pc_.end()) {
+                tdata->error =
+                    "Unknown pc after ir2trace: did ir2trace insert new instr?";
+                return false;
+            }
+            saved_decode_pc = tdata->syscall_pc_to_decode_pc_[instr_pc].first;
+            if (!append_encoding(tdata, saved_decode_pc, entry.size, buf,
+                                 entries_with_encodings))
+                return false;
+        }
+        *buf = entry;
+        ++buf;
     }
-
-    if (!tdata->out_file->write(reinterpret_cast<const char *>(entries.data()),
-                                sizeof(trace_entry_t) * entries.size())) {
-        tdata->error = "Failed to write to output file";
-        return false;
+    if (buf != entries_with_encodings) {
+        if (!write(tdata, entries_with_encodings, buf, &saved_decode_pc, 1)) {
+            return false;
+        }
     }
+    return true;
+}
 
+bool
+raw2trace_t::track_syscall_pt_encodings(raw2trace_thread_data_t *tdata,
+                                        instrlist_t *ilist)
+{
+    instr_t *instr = instrlist_first(ilist);
+    while (instr != NULL) {
+        app_pc orig_pc = instr_get_app_pc(instr);
+        int instr_len = instr_length(dcontext_, instr);
+        if (tdata->syscall_instr_encodings_.empty() ||
+            tdata->syscall_next_encoding_offset_ + instr_len >=
+                SYSCALL_PT_ENCODING_BUF_SIZE) {
+            // We allocate new memory to store kernel instruction encodings in
```

Comment:
>  I see pt2ir is making a copy of the encoding for every instr today

You mean the following? https://github.com/DynamoRIO/dynamorio/blob/cd5521ce48238af7384c79716891640161131c3d/clients/drcachesim/drpt2trace/pt2ir.cpp#L387

That's pt2ir decoding the insn raw bits returned by libipt to create the instr_t used here (we need to create instr_t because the memory address reconstruction logic in ir2trace needs the full instr_t). libipt internally sets those raw bits; insn is discarded after we create the instr_t.

I guess getting the bits from kcore is also a possibility, but I wasn't sure if libipt is doing any other magic in its implementation. But since we have the raw bits given by libipt anyway, I thought there's value in just using that instead of going back into the kcore ourselves.

Reply:
I think it should be possible to reuse the insn bits here instead of encoding the instr again. Would still be better than getting them from the kcore again ourselves.